### ISSUE TITLE
Task List

### ISSUE THREAD

**Issue by PulseBeat02** [PulseBeat02/mcav#3]
# **Task List:**

too lazy to update tasks on Github. To do list can be found in Discord.

**Comment by mishal321** [comment]
Hey, pulse! Im interested in this..  How do you exactly implement this into a server...?

**Comment by PulseBeat02** [comment]
> Hey, pulse! Im interested in this.. How do you exactly implement this into a server...?

Hi mishal! Thank you for taking the time to take a look at this repository. The library is unfinished, but I can give you some information about how I implement this into a server.

Due to Java's slow code rate, we must (and should) use native libraries written in C/C++ for video playback because using Java would be just too slow. One of the ways to do this is by using [VLC Media Player](https://www.videolan.org/vlc/), or more specifically the binaries. There is another repository called [VLCJ](https://github.com/caprica/vlcj) which creates VLC bindings in Java and allows certain callbacks to be handled. I use this technique as a way to fetch the frames from a video file.

Now obviously, this isn't as clean as cutting butter. There are some issues (such as VLCJ needing a VLC installation beforehand, but we can get to these later).

As for the main example, I use mainly Maps to render the video pixels. I display such pixels onto the map and send packets to the players who are able to view the packets. As an example, look at this NMS implementation [here](https://github.com/PulseBeat02/MinecraftMediaLibrary/blob/master/v1_16_R3/src/main/java/com/github/pulsebeat02/minecraftmedialibrary/nms/impl/v1_16_R3/NMSMapPacketIntercepter.java), where I set custom packets certain attributes in order to send the proper data to the clients.

As for rendering the pixels on the map, it is a bit harder. I could use a standard technique where I just try to display them onto the map, but this isn't possible because Minecraft has a limited palette (which was somewhere around 80 colors). That means I have to use a dithering algorithm to dither each of the frames of the picture. This allows for a better overall quality image. My dithering algorithms implementations can be found [here](https://github.com/PulseBeat02/MinecraftMediaLibrary/tree/master/MinecraftMediaLibrary/src/main/java/com/github/pulsebeat02/minecraftmedialibrary/video/dither). The fastest dithering algorithm while producing decent results would be Filter Lite dithering.

The repository is very big, but I documented a ton of the code. If you have any questions, however, feel free to just ask me here!
